### PR TITLE
feat: launchpad client

### PR DESCRIPTION
**Pull request by lengau**
This client is a wrapper around launchpadlib that provides a unified interface. It's heavily inspired by the Launchpad wrapper used in snapcraft's most recent remote build, but with tweaks to make it more generic.

- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

### CODE REVIEW

File: craft_application/launchpad/models/recipe.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+    name: str
+    """The name of the recipe."""
+    owner_name: str
+    """The name of the owner."""
+
+    _resource_types = RecipeType
+    _attr_map = {"owner_name": "owner.name"}
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}.get(name={self.name!r}, owner={self.owner_name!r})"
```
Comment (syu-w) [inline craft_application/launchpad/models/recipe.py:82]: I am not sure if this is suppose to be formatted to`"NAME.get(name=N, owner=O)"` or you want to call `NAME.get(name=N, owner=O)` and get the result.

File: craft_application/launchpad/models/recipe.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+            "archive": archive,
+            "pocket": pocket.value,
+        }
+        if channels:
+            request_build_kwargs["channels"] = channels
+        build_request = self._obj.requestBuilds(**request_build_kwargs)
+        sleep_time = 0.5
+        while build_request.status == "Pending":
+            # Check to see if we've run out of time.
+            if deadline is not None and time.monotonic_ns() >= deadline:
```
Comment (syu-w) [inline craft_application/launchpad/models/recipe.py:127]: This needs to use nanoseconds?

File: craft_application/launchpad/models/recipe.py (hunk 1 above)
Comment (lengau) [reply craft_application/launchpad/models/recipe.py:82]: It's a bit weird and I'll add a comment, but I'm intending to format it the first way.

File: craft_application/launchpad/models/recipe.py (hunk 2 above)
Comment (lengau) [reply craft_application/launchpad/models/recipe.py:127]: No, but `time.monotonic()` is basically just `time.monotonic_ns() / 10**9` and I prefer working with integers.

File: craft_application/launchpad/launchpad.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        """Get a recipe.
+
+        :param type_: The type of recipe to get.
+        :param owner: The owner of the recipe.
+        :param name: The name of the recipe.
+        :param project: (Optional) The project to which the recipe is attached
+        :returns: A Recipe object of the appropriate type.
+        :raises: ValueError if the type requested is invalid.
+
+        A charm recipe requires a project. If none is given, the first charm matching
```
Comment (tigarmo) [inline craft_application/launchpad/launchpad.py:109]: should this bit of the docstring be made more generic?

File: craft_application/services/lifecycle.py [hunk 4]
```diff
@@ -124,7 +124,7 @@ class LifecycleService(base.ProjectService):
         LifecycleManager on initialisation.
     """
 
-    def __init__(  # noqa: PLR0913 (too many arguments)
+    def __init__(  # (too many arguments)
```
Comment (tigarmo) [inline craft_application/services/lifecycle.py:127]: clean these up maybe?

**Review by tigarmo — APPROVED** [review]
Thanks. This is a lot of code but it seems overall fine and I think we can take it out for a spin

File: craft_application/launchpad/launchpad.py (hunk 3 above)
Comment (lengau) [reply craft_application/launchpad/launchpad.py:109]: It's noting about a special case, but I'll make that more obvious.

File: craft_application/launchpad/models/recipe.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    def new(
+        cls,
+        lp: Launchpad,
+        name: str,
+        owner: str,
+        *,
+        architectures: Collection[str] | None = None,
+        description: str | None = None,
+        project: str | None = None,
+        # Repository. Must be either a git repository or a Bazaar branch but not both.
```
Comment (mr-cal) [inline craft_application/launchpad/models/recipe.py:178]: Would prefer docstrings but I'm OK with this style.

File: craft_application/launchpad/models/base.py [hunk 6]
```diff
@@ -159,7 +175,7 @@ def __setattr__(self, key: str, value: Any) -> None:  # noqa: ANN401
             )
 
     def __repr__(self) -> str:
-        """String of this Launchpad object."""
+        # """Get a machine-readable string of this Launchpad object."""
```
Comment (mr-cal) [inline craft_application/launchpad/models/base.py:178]:
```diff
-# """Get a machine-readable string of this Launchpad object."""
+"""Get a machine-readable string of this Launchpad object."""
```

**Review by mr-cal — APPROVED** [review]
Nice, lots of code but no code smells (aside from pyright + launchpad, which is unavoidable).

Let 'er rip!

File: craft_application/launchpad/launchpad.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/launchpad/launchpad.py]: I've updated this based on Tiago's comment to clarify :-)

File: craft_application/launchpad/models/base.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+from typing import TYPE_CHECKING
+
+import lazr.restfulclient.errors
+from lazr.restfulclient.resource import Entry  # type: ignore[import-untyped]
+from typing_extensions import Any, Self
+
+from .. import errors, util
+
+if TYPE_CHECKING:
+    from ...launchpad import Launchpad
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/launchpad/models/base.py:43]: Two levels up. Pycharm did this automatically, but really it's probably better to go one level up. (Adjusted!)

File: craft_application/launchpad/models/recipe.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+class RecipeType(enum.Enum):
+    """The type of recipe."""
+
+    SNAP = "snap"
+    CHARM = "charm_recipe"
+
+
+class BuildChannels(TypedDict, total=False):
+    """Typed dictionary for the snap channels for recipe builds."""
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/launchpad/models/recipe.py:56]: I'm conflicted, but in the end I just made a longer docstring to explain better.

File: craft_application/launchpad/models/recipe.py (hunk 5 above)
Comment (lengau) [reply craft_application/launchpad/models/recipe.py:178]: I've added some docstrings but kept these comments as it helps developers group the large number of parameters.

**Pull request merged**